File: server/domain/PageType.py
from enum import Enum


class PageType(Enum):
    twitter = "twitter"
    hoyolab = "hoyolab"
    hoyoLink = "hoyoLink"
    instagram = "instagram"

    @staticmethod
    def from_str(host_name: str):
        match host_name:
            case "x.com" | "twitter.com":
                return PageType.twitter
            case "www.hoyolab.com":
                return PageType.hoyolab
            case "hoyo.link":
                return PageType.hoyoLink
            case "www.instagram.com":
                return PageType.instagram

    def url(self, source_id: str):
        match self:
            case PageType.twitter:
                return "https://x.com/" + source_id
            case PageType.hoyolab:
                return "https://www.hoyolab.com/article/" + source_id
            case PageType.instagram:
                return "https://www.instagram.com/p/" + source_id

File: internal/model/entity/pageType/pageType.go
package pageType

import "errors"

type PageType string

const (
    Twitter   PageType = "twitter"
    HoyoLab   PageType = "hoyolab"
    HoyoLink  PageType = "hoyolink"
    Instagram PageType = "instagram"
    Null      PageType = ""
)

func FromHost(hostName string) (PageType, error) {
    switch hostName {
    case "x.com":
        fallthrough
    case "twitter.com":
        return Twitter, nil
    case "www.instagram.com":
        return Instagram, nil
    case "www.hoyolab.com":
        return HoyoLab, nil
    case "hoyo.link":
        return HoyoLink, nil
    }
    return Null, errors.New("invalid host name")
}

func (p PageType) Url(souceId string) string {
    switch p {
    case Twitter:
        return "https://x.com/" + souceId
    case HoyoLab:
        return "https://www.hoyolab.com/article/" + souceId
    case Instagram:
        return "https://www.instagram.com/p/" + souceId
    default:
        return ""
    }
}
